src/web-extensions.c

```diff
@@ -0,0 +1,861 @@
+  g_autoptr(GError) error = NULL;
+  Session *session;
+
+  session = (Session *)web_extensions_session_new (arg_options, call, connection, &error);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+  if (!session_export (session, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      session_close (session, FALSE);
+      return TRUE;
+    }
+  session_register (session);
+
+  xdp_dbus_web_extensions_complete_create_session (object, invocation, session->id);
+
+  return TRUE;
+}
+
+static void
+on_host_exited (GPid pid,
+                gint status,
+                gpointer user_data)
+{
+  Session *session = user_data;
+  WebExtensionsSession *web_extensions_session = (WebExtensionsSession *)session;
+
+  SESSION_AUTOLOCK (session);
+  web_extensions_session->child_pid = -1;
+  web_extensions_session->child_watch_id = 0;
+  session_close (session, TRUE);
+}
+
+static gboolean
+array_contains (JsonArray *array,
+                const char *value)
+{
+  guint length, i;
+
+  if (array == NULL)
+    return FALSE;
+
+  length = json_array_get_length (array);
+  for (i = 0; i < length; i++)
+    {
+      const char *element = json_array_get_string_element (array, i);
+      if (g_strcmp0 (element, value) == 0)
+        return TRUE;
+    }
+  return FALSE;
+}
+
+static gboolean
+is_valid_name (const char *name)
+{
+  /* This regexp comes from the Mozilla documentation on valid native
+     messaging host names:
+
+     https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#native_messaging_manifests
+
+     That is, one or more dot-separated groups composed of
+     alphanumeric characters and underscores.
+  */
+  return g_regex_match_simple ("^\\w+(\\.\\w+)*$", name, 0, 0);
+}
+
+static GStrv
+get_manifest_search_path (WebExtensionsSessionMode mode)
+{
+  /* IMPORTANT:
+     The safety model depends on the inability of the sandboxed
+     browser to write to the search locations specified below.
+
+     As this portal allows browser extensions to run a native
+     messaging application outside of the sandbox through the portal,
+     the sandboxing mechanism must ensure that these locations are
+     inaccessible to the browser.  If the locations are both readable
+     AND writable by the sandboxed browser, then a vulnerability
+     resulting in a file writing primitive within the sandbox could
+     result in arbitrary code execution outside of the sandbox through
+     the portal.
+
+     For example, the Firefox Snap package meets this criterion,
+     because all strictly confined Snap packages (including Firefox)
+     are prohibited by AppArmor from accessing most directories and
+     files in the user's home directory, except where explicitly
+     specified, for instance using the 'personal-files' interface.
+     https://snapcraft.io/docs/personal-files-interface
```

Comment:
```suggestion
     https://snapcraft.io/docs/personal-files-interface
   The configuration of the Firefox snap package is at
    https://github.com/canonical/firefox-snap/blob/stable/snapcraft.yaml
```

Let's also link the location of the snapcraft file for Firefox snap so that readers can easily verify the claims from this comment.